Realm/Realm.SourceGenerator/InfoClasses.cs

```diff
@@ -0,0 +1,305 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Realms.SourceGenerator
+{
+    internal record ClassInfo
+    {
+        public string Name { get; set; }
+
+        public bool IsEmbedded { get; set; }
+
+        public string MapTo { get; set; }
+
+        public string Namespace { get; set; }
+
+        public Accessibility Accessibility { get; set; }
+
+        public ITypeSymbol TypeSymbol { get; set; }
+
+        public List<PropertyInfo> Properties { get; } = new();
+
+        public List<Diagnostic> Diagnostics { get; } = new();
+
+        public List<string> Usings { get; } = new();
+
+        public List<EnclosingClassInfo> EnclosingClasses { get; } = new();
+
+        public bool HasParameterlessConstructor { get; set; }
+
+        public PropertyInfo PrimaryKey => Properties.FirstOrDefault(p => p.IsPrimaryKey);
+    }
+
+    internal record EnclosingClassInfo
+    {
+        public string Name { get; set; }
+
+        public Accessibility Accessibility { get; set; }
+    }
+
+    internal record PropertyInfo
+    {
+        public string Name { get; set; }
+
+        public bool IsIndexed { get; set; }
+
+        public bool IsRequired { get; set; }
+
+        public bool IsPrimaryKey { get; set; }
+
+        public string MapTo { get; set; }
+
+        public string Backlink { get; set; }
+
+        public PropertyTypeInfo TypeInfo { get; set; }
+
+        public Accessibility Accessibility { get; set; }
+
+        public string Initializer { get; set; }
+    }
+
+    internal abstract record PropertyTypeInfo
+    {
+        private static HashSet<SimpleTypeEnum?> _indexableTypes = new()
+        {
+            SimpleTypeEnum.Int,
+            SimpleTypeEnum.Bool,
+            SimpleTypeEnum.String,
+            SimpleTypeEnum.ObjectId,
+            SimpleTypeEnum.Guid,
+            SimpleTypeEnum.Date,
+        };
+
+        private static HashSet<SimpleTypeEnum?> _primaryKeyTypes = new()
+        {
+            SimpleTypeEnum.Int,
+            SimpleTypeEnum.String,
+            SimpleTypeEnum.ObjectId,
+            SimpleTypeEnum.Guid,
+        };
+
+        private static HashSet<SimpleTypeEnum?> _requiredTypes = new()
+        {
+            SimpleTypeEnum.String,
+            SimpleTypeEnum.Data,
+        };
+
+        public virtual SimpleTypeEnum? SimpleType { get; set; }
+
+        public virtual CollectionTypeEnum? CollectionType { get; set; }
+
+        public virtual bool IsRealmInteger { get; set; }
+
+        public virtual bool IsIQueryable { get; set; }
```

Comment:
I can understand the feeling, but I think most of the time IQueryable needs to be treated quite differently than the other two, so this could make the code even more difficult to read.